File: app/src/main/java/org/blitzortung/android/map/overlay/StrikeOverlayItem.java
package org.blitzortung.android.map.overlay;

import android.graphics.Point;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.shapes.Shape;
import android.location.Location;
import com.google.android.maps.GeoPoint;
import com.google.android.maps.OverlayItem;
import com.google.android.maps.Projection;
import org.blitzortung.android.data.Coordsys;
import org.blitzortung.android.data.beans.StrikeAbstract;
import org.blitzortung.android.data.beans.RasterParameters;
import org.blitzortung.android.data.beans.Strike;

public class StrikeOverlayItem extends OverlayItem implements Strike {

	private final long timestamp;
	
	private final int multiplicity;

    private static final Point center = new Point();

    private static final Point topLeft = new Point();

    private static final Point bottomRight = new Point();

	public StrikeOverlayItem(StrikeAbstract strike) {
		super(Coordsys.toMapCoords(strike.getLongitude(), strike.getLatitude()), "", "");
        super.setMarker(new ShapeDrawable());

		timestamp = strike.getTimestamp();
		multiplicity = strike.getMultiplicity();
	}
	
    @Override
	public long getTimestamp() {
		return timestamp;
	}

    @Override
    public Location getLocation(Location location) {
        final GeoPoint point = getPoint();
        location.setLongitude(point.getLongitudeE6() / 1e6);
        location.setLatitude(point.getLatitudeE6() / 1e6);
        return location;
    }

    @Override
    public int getMultiplicity() {
		return multiplicity;
	}

    @Override
    public void setMarker(Drawable drawable) {
        throw new IllegalStateException("cannot overwrite marker of strike overlay item");
    }

    private ShapeDrawable getDrawable() {
        return (ShapeDrawable)getMarker(0);
    }
    
    public Shape getShape() {
        return getDrawable().getShape();
    }
    
    public void setShape(Shape shape) {
        getDrawable().setShape(shape);
    }

    public void updateShape(RasterParameters rasterParameters, Projection projection, int color, int textColor, int zoomLevel) {
        Shape shape = getShape();
        if (rasterParameters != null) {
            if (shape == null) {
                shape = new RasterShape();
            }
            RasterShape rasterShape = (RasterShape)shape;

            float lon_delta = rasterParameters.getLongitudeDelta() / 2.0f * 1e6f;
            float lat_delta = rasterParameters.getLatitudeDelta() / 2.0f * 1e6f;
            GeoPoint geoPoint = getPoint();
            projection.toPixels(geoPoint, center);
            projection.toPixels(new GeoPoint(
                    (int) (geoPoint.getLatitudeE6() + lat_delta),
                    (int) (geoPoint.getLongitudeE6() - lon_delta)), topLeft);
            projection.toPixels(new GeoPoint(
                    (int) (geoPoint.getLatitudeE6() - lat_delta),
                    (int) (geoPoint.getLongitudeE6() + lon_delta)), bottomRight);
            topLeft.offset(-center.x, -center.y);
            bottomRight.offset(-center.x, -center.y);
            rasterShape.update(topLeft, bottomRight, color, getMultiplicity(), textColor);
        } else {
            if (shape == null) {
                shape = new StrikeShape();
            }
            StrikeShape strikeShape = (StrikeShape) shape;
            strikeShape.update(zoomLevel + 1, color);

        }
        setShape(shape);
    }
}

File: app/src/main/java/org/blitzortung/android/map/overlay/StrikeOverlayItem.kt
/*

   Copyright 2015 Andreas Würl

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

*/

package org.blitzortung.android.map.overlay

import android.graphics.Point
import android.graphics.drawable.Drawable
import android.graphics.drawable.ShapeDrawable
import android.graphics.drawable.shapes.Shape
import com.google.android.maps.GeoPoint
import com.google.android.maps.OverlayItem
import com.google.android.maps.Projection
import org.blitzortung.android.data.Coordsys
import org.blitzortung.android.data.beans.RasterParameters
import org.blitzortung.android.data.beans.Strike

class StrikeOverlayItem(strike: Strike) : OverlayItem(Coordsys.toMapCoords(strike.longitude, strike.latitude), "", ""), Strike {
    override val timestamp: Long
    override val multiplicity: Int

    init {
        super.setMarker(ShapeDrawable())

        timestamp = strike.timestamp
        multiplicity = strike.multiplicity
    }

    override fun setMarker(drawable: Drawable?) {
        throw IllegalStateException("cannot overwrite marker of strike overlay item")
    }

    private val drawable: ShapeDrawable
        get() {
            return getMarker(0) as ShapeDrawable
        }

    var shape: Shape?
        get() {
            return drawable.shape
        }
        set(shape) {
            drawable.shape = shape
        }

    fun updateShape(rasterParameters: RasterParameters?, projection: Projection, color: Int, textColor: Int, zoomLevel: Int) {
        var shape: Shape? = shape
        if (rasterParameters != null) {
            if (shape == null && shape !is RasterShape) {
                shape = RasterShape()
            }

            val lon_delta = rasterParameters.longitudeDelta / 2.0f * 1e6f
            val lat_delta = rasterParameters.latitudeDelta / 2.0f * 1e6f
            val geoPoint = point
            projection.toPixels(geoPoint, center)
            projection.toPixels(GeoPoint(
                    (geoPoint.latitudeE6 + lat_delta).toInt(),
                    (geoPoint.longitudeE6 - lon_delta).toInt()), topLeft)
            projection.toPixels(GeoPoint(
                    (geoPoint.latitudeE6 - lat_delta).toInt(),
                    (geoPoint.longitudeE6 + lon_delta).toInt()), bottomRight)
            topLeft.offset(-center.x, -center.y)
            bottomRight.offset(-center.x, -center.y)
            if (shape is RasterShape) {
                shape.update(topLeft, bottomRight, color, multiplicity, textColor)
            }
        } else {
            if (shape == null) {
                shape = StrikeShape()
            }
            if (shape is StrikeShape) {
                shape.update((zoomLevel + 1).toFloat(), color)
            }
        }
        this.shape = shape
    }

    override val longitude: Float
        get() {
            return point.longitudeE6 / 1e6f
        }

    override val latitude: Float
        get() {
            return point.latitudeE6 / 1e6f
        }

    companion object {
        private val center = Point()
        private val topLeft = Point()
        private val bottomRight = Point()
    }
}
